tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/observability/metrics/MetricsTestCase.java

```diff
@@ -0,0 +1,312 @@
+        filteredMetrics.addAllPolledGauges(filterByTag(metrics.getPolledGauges(), key, value));
+        return filteredMetrics;
+    }
+
+    private <M extends MockMetric> List<M> filterByTag(List<M> metrics, String key, String value) {
+        return metrics.stream()
+                .filter(metric -> {
+                    Optional<Tag> tag = metric.getId().getTags().stream()
+                            .filter(t -> Objects.equals(t.getKey(), key))
+                            .findFirst();
+                    return (tag.isEmpty() && value == null)
+                            || (tag.isPresent() && Objects.equals(tag.get().getValue(), value));
+                })
+                .collect(Collectors.toList());
+    }
+
+    private void assertFunctionMetrics(Metrics allMetrics, String invocationPosition, long invocationCount,
+                                       Tag ...additionalTags) {
+        Metrics functionMetrics = filterByTag(allMetrics, "src.position", invocationPosition);
+        Set<Tag> tags = new HashSet<>(Arrays.asList(additionalTags));
+        tags.add(Tag.of("src.module", MODULE_ID));
+        tags.add(Tag.of("src.position", invocationPosition));
+
+        Assert.assertEquals(functionMetrics.getCounters().stream()
+                        .map(gauge -> gauge.getId().getName())
+                        .collect(Collectors.toSet()),
+                new HashSet<>(Arrays.asList("requests_total", "response_time_nanoseconds_total")));
+        Assert.assertEquals(functionMetrics.getCounters().size(), 2);
+        functionMetrics.getCounters().forEach(counter -> {
+            Assert.assertEquals(counter.getId().getTags(), tags);
+            if (Objects.equals(counter.getId().getName(), "requests_total")) {
+                Assert.assertEquals(counter.getValue(), invocationCount);
+            } else if (Objects.equals(counter.getId().getName(), "response_time_nanoseconds_total")) {
+                if (invocationCount > 0) {
+                    Assert.assertTrue(counter.getValue() > 0,
+                            "response_time_nanoseconds_total expected to be greater than 0, but found "
+                            + counter.getValue());
+                } else {
+                    Assert.assertEquals(counter.getValue(), 0);
+                }
+            } else {
+                Assert.fail("Unexpected metric " + counter.getId().getName());
+            }
+        });
+
+        Assert.assertEquals(functionMetrics.getGauges().stream()
+                        .map(gauge -> gauge.getId().getName())
+                        .collect(Collectors.toSet()),
+                new HashSet<>(Arrays.asList("response_time_seconds", "inprogress_requests")));
+        Assert.assertEquals(functionMetrics.getGauges().size(), 2);
+        functionMetrics.getGauges().forEach(gauge -> {
+            if (Objects.equals(gauge.getId().getName(), "response_time_seconds")) {
```

Comment:
This if condition block seems to be too big and should be refactored into multiple methods for better readability.

Reply:
Fixed in f2a17a0d2edd186d40174bd5bea21f9551b363a6